### ISSUE TITLE
Reconnection behavior when client loses connection / sleeps

### ISSUE THREAD

**Issue by Ezekiel-DA** [aws/aws-iot-device-sdk-js-v2#72]
Confirm by changing [ ] to [x] below to ensure that it's a bug:
- [ x] I've searched for [previous similar issues](https://github.com/aws/aws-iot-device-sdk-js-v2/issues) and didn't find any solution
  
**Describe the bug**
Hi,

I'm trying to use the basic pub/sub mechanism as per the sample. I have slightly modified said sample. as seen below. The relevant changes are:
* using Javascript directly
* simplified a few things I didn't need
* added a naive attempt at handling disconnections, simply by reconnecting every 5 minutes

Here's my issue: I have a situation where the client fails to keep the connection alive (in this specific, quite simply because the computer running this is in sleep mode).

When this occurs, the MQTT broker in AWS IoT Core disconnects the client, as shown in its logs (at least that's how I'm interpreting the `"clientInitiatedDisconnect": false` below):
```
{
  "clientId": "desktop-pc-controller",
  "timestamp": 1594914638289,
  "eventType": "disconnected",
  "clientInitiatedDisconnect": false,
  "sessionIdentifier": "9d302d02-a6f8-44b1-92a1-0784f0db8072",
  "principalIdentifier": "487854dcfaf7ac1aba3a30326287d36827060c24a07e5f2730aaee3ba5f67b5f",
  "disconnectReason": "MQTT_KEEP_ALIVE_TIMEOUT",
  "versionNumber": 2
}
```

This makes sense to me so far.

What doesn't is that when my app regains network connection / resumes from sleep, it mistakenly believes the connection to still be okay. No event or exception occurs that I can see, so I have no way to know that my subscription will never receive messages again.

Worse, when my naive attempt at handling this occurs (a simple reconnect every 5 minutes, as mentioned above), the same thing occurs as when we are correctly still connected: a `AWS_ERROR_MQTT_ALREADY_CONNECTED` error is thrown.

This behavior seems like a bug to me? Unless there is some sort of best practice I am not following that would allow me to detect a broker initiated disconnection has occurred and restore the connection?

**SDK version number**
1.2.2

**Platform/OS/Hardware/Device**
Windows 10 / PC / NodeJS v12.18.2

**To Reproduce (observed behavior)**
```
const {mqtt, io, iot} = require('aws-crt')
const {TextDecoder} = require('util')

const endpoint = <removed>
const ca_file = <removed>
const cert = <removed>
const key = <removed>
const client_id = 'desktop-pc-controller'

const initMQTTConnection = async () => {
    const client_bootstrap = new io.ClientBootstrap();

    let config_builder = null;
    config_builder = iot.AwsIotMqttConnectionConfigBuilder.new_mtls_builder_from_path(cert, key);  
    config_builder.with_certificate_authority_from_path('./', ca_file);
    
    config_builder.with_clean_session(true);
    config_builder.with_client_id(client_id);  
    config_builder.with_endpoint(endpoint);

    const config = config_builder.build();
    const client = new mqtt.MqttClient(client_bootstrap);
    const connection = client.new_connection(config);

    await connection.connect()
    return connection
}

async function main() {
    try {
        let connection
        
        setInterval(() => {
            console.log('Alive...');
            if (connection) {
                connection.connect().then(() => console.log('reconnected')).catch((e) => {
                    if (e && e.message && /AWS_ERROR_MQTT_ALREADY_CONNECTED/.test(e.message))
                        return // catch and burry; we probably shouldn't be reconnecting all the time but I don't know how to work around Sleep induced disconnections yet...
                    else { // caught something real, rethrow
                        console.log(e)
                        throw e
                    }
                })
            }
        }, 5* 60 * 1000)

        connection = await initMQTTConnection()
        connection.addListener('close', (...args) => console.log(`Caught error event; ${args}`))
        connection.addListener('close', (...args) => console.log(`Caught close event; ${args}`))
        connection.subscribe('screen_control', mqtt.QoS.AtLeastOnce, () => console.log('handling of my message ommited for brevity'))
    } catch (e) {
        // assume we caught a connection error and reconnect
        console.log(`Caught error: ${e}\nAborting.`)
        process.exit(0)
    }
}

main()
```

**Expected behavior**
Connection should either auto-restore, or some sort of "disconnected by broker" event should be emited that I could listen to and reconnect.

Thanks in advance for your help!

**Comment by jmklix** [comment]
Hello Ezekiel-DA and sorry for the late reply. Re-connections are handled internally the the sdk, so you don't need to include your own re-connections. That is why you are getting the AWS_ERROR_MQTT_ALREADY_CONNECTED error. Please let us know if you have any other problems with the js sdk.

**Comment by Ezekiel-DA** [comment]
@jmklix Hi Joseph,

thank you for the response!

Unfortunately what you describe is exactly my issue. To clarify what appears to be happening: I believe that if the connection is dropped by the broker (while the client is sleeping and cannot learn about it), the client never realizes this. Attempts to reconnect are met with the AWS_ERROR_MQTT_ALREADY_CONNECTED, but the broker does not have this client in its list anymore and messages are not sent to it.

Does that seem possible? Am I missing some simple way to handle this scenario?

**Comment by TingDaoK** [comment]
Hi @Ezekiel-DA,

It's possible that the client side still think it's connected after the sleep. What you need to do is not calling connect again. Instead, if keep-alive timeout happens, the client will try to ping broker to find out the connection is still alive or not. So, in this case, you can set your keep-alive time out to 5 mins by
```
config.keep_alive = 5 * 60; //secs -> 5 mins.
```
SDK will automatically detect the connection lost once timeout happens and try to reconnect by itself.

To track what's happening, you can set interrupt&resume handler like
```
connection.on('interrupt', (err) => { console.log(`Interrupt with ${err}`); });
connection.on('resume', (return_code, session_present) => { console.log(`Resume with ${return_code} ${session_present}`); }); 

```

But, we had bug fixed for the handlers, and the release has not published yet. We will cut new release soon. Make sure updating to the latest version.

**Comment by rccarper** [comment]
Hello, the bug fix mentioned above for interrupt/resume behavior has been released in v1.3.0.
